Enable multiple secrets dirs

makukha opened the pull request:
* Allow `secrets_dir` to be a tuple or list of directories

Closes:
* #3
* #371

makukha commented:
Ready for review!

makukha commented:
@hramezani it seems that I don't have permissions to ask for review, but this PR is probably ready.

hramezani commented:
> @hramezani it seems that I don't have permissions to ask for review, but this PR is probably ready.

Thanks @makukha for the PR. I will review it later

hramezani commented on tests/test_settings.py:
```diff
@@ -1416,6 +1416,30 @@ class Settings(BaseSettings):
+    (d2 / 'foo2').write_text('foo2_dir2_secret_value_str')
+    (d2 / 'foo3').write_text('foo3_dir2_secret_value_str')
+
+    class Settings(BaseSettings):
+        foo1: str
+        foo2: str
+        foo3: str
+
+        model_config = SettingsConfigDict(secrets_dir=(d1, d2))
+
+    assert Settings().model_dump() == {
+        'foo1': 'foo1_dir1_secret_value_str',
+        'foo2': 'foo2_dir2_secret_value_str',  # dir2 takes priority
+        'foo3': 'foo3_dir2_secret_value_str',
+    }
```
please add another model with reverse `secrets_dir`

hramezani commented on pydantic_settings/sources.py:
```diff
@@ -594,14 +594,17 @@ def __call__(self) -> dict[str, Any]:
-            raise SettingsError(f'secrets_dir must reference a directory, not a {path_type_label(self.secrets_path)}')
+        for path in self.secrets_paths:
+            if not path.is_dir():
+                raise SettingsError(f'secrets_dir must reference a directory, not a {path_type_label(path)}')
```
There is a test for this error that tests single dir. please add a test for multiple directories as well.

hramezani commented on pydantic_settings/sources.py:
```diff
@@ -594,14 +594,17 @@ def __call__(self) -> dict[str, Any]:
-            return secrets
+        for path in self.secrets_paths:
+            if not path.exists():
+                warnings.warn(f'directory "{path}" does not exist')
```
There is a test for this warning that tests single dir. please add a test for multiple directories as well.

hramezani commented on pydantic_settings/sources.py:
```diff
@@ -639,18 +642,20 @@ def get_field_value(self, field: FieldInfo, field_name: str) -> tuple[Any, str,
+                if path.is_file():
+                    return path.read_text().strip(), field_key, value_is_complex
+                else:
+                    warnings.warn(
```
There is a test for this warning that tests single dir. please add a test for multiple directories as well.

makukha commented:
@hramezani adding tests for warnings and errors was a good idea, fixed a bug. Also, I updated docs describing behaviour when secrets dirs can be missing or can be files.

hramezani commented on tests/test_settings.py:
```diff
@@ -1544,6 +1593,37 @@ class Settings(BaseSettings):
+    (d1 := tmp_path / 'dir1').mkdir()
+    (d1 / 'foo').write_text('secret_value')
+
+    with pytest.warns() as record:
```
can't you use `with pytest.warns(UserWarning, match=...)` for this test and the others?

makukha replied:
Fixed for two cases where a single warning is emitted. For other cases, we have multiple warnings, and need a `record` object from `pytest.warns() as record` to check both. I wouldn't further simplify those cases.

hramezani commented:
Thanks @makukha

makukha commented:
@hramezani thank you for supporting this awesome package!
